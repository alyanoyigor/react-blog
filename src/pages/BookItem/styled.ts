import styledEngine from '@mui/styled-engine';
import { IconButton } from '@mui/material';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';

export const StyledButton = styledEngine(IconButton)`
  color: #fafafa;
  border-radius: 10px;
  display: flex;
  align-items: center;
  font-size: 20px;

  & .icon {
    color: #ffffff;
  }

  &.active {
    background-color: rgba(255, 255, 255, 0.08);

    &:hover {
      background-color: rgba(255, 255, 255, 0.12);
    }
  }

  &:hover {
    background-color: rgba(255, 255, 255, 0.04);
  }
`;

export const StyledBackIcon = styledEngine(ArrowBackIcon)`
  color: #ffffff;
`;
