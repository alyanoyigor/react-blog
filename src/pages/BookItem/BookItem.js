import React from 'react';
import moment from 'moment';
import { useNavigate, useParams } from 'react-router-dom';
import { getBookItem } from '../../api/books';
import { Error } from '../../components/Error';
import { useAxios } from '../../hooks';
import { Preloader } from '../../components/Preloader';
import { Box, Typography } from '@mui/material';
import { StyledBackIcon, StyledButton } from './styled';

export const BookItem = () => {
  const { bookId } = useParams();
  const {
    error,
    data: bookData,
    loading,
  } = useAxios(() => getBookItem(bookId), true);
  const navigate = useNavigate();

  return (
    <>
      {loading && !error && <Preloader />}
      {bookData && !loading && !error && (
        <>
          <StyledButton onClick={() => navigate(-1)}>
            <StyledBackIcon />
            <span>Back</span>
          </StyledButton>
          <Box display="flex" flexDirection="column" alignItems="center" mt={2}>
            <Typography mb={1} variant="h4" component="h1">
              {bookData.title}
            </Typography>
            <Typography variant="subtitle1" component="p">
              {moment(bookData.publishDate).format('MMM DD, YYYY')}
            </Typography>
            <Typography>{bookData.pageCount} pages</Typography>
            <Typography>{bookData.description}</Typography>
            <Typography textAlign="center">{bookData.excerpt}</Typography>
          </Box>
        </>
      )}
      {error && !loading && <Error>{error}</Error>}
    </>
  );
};
