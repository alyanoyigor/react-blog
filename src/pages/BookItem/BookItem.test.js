import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { act, screen, waitForElementToBeRemoved } from '@testing-library/react';
import { renderWithRedux } from '../../test';
import { renderWithProviders } from '../../test';
import { BookItem } from './BookItem';

const response = {
  _id: '62de6a961301da01ad8e4081',
  title: '1',
  description: '1',
  date: '2022-07-25T09:28:44.802Z',
  pages: 1,
  excerpt: '1',
  createdAt: '2022-07-25T10:04:06.198Z',
  updatedAt: '2022-07-25T10:07:38.841Z',
  __v: 0,
};

const handlers = [
  rest.get('/books/:bookId', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json({ data: response }), ctx.delay(200));
  }),
];
const server = setupServer(...handlers);

describe('BookItem', () => {
  const bookItemOptions = {
    route: `/books/${response._id}`,
    params: '/books/:bookId',
  };

  beforeAll(() => server.listen());

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => server.close());

  test('should render', () => {
    const { container } = renderWithProviders(<BookItem />, bookItemOptions);
    expect(container).toBeInTheDocument();
  });

  test('should call thunk inside useEffect after mounting', async () => {
    act(() => {
      renderWithProviders(<BookItem />, bookItemOptions);
    });

    await waitForElementToBeRemoved(() => screen.getByTestId('preloader'));

    const title = await screen.findByTestId('book-title');
    expect(title).toBeInTheDocument();
    expect(title.textContent).toEqual(response.title);
  });

  // test('should reset data after unmounting', () => {});

  // test('should show preloader when data is loading', () => {});

  // test('should match to snapshot with success response data', () => {});

  // test('should match to snapshot with error response data', () => {});

  // test('should move to previous path after back button clicking', () => {});
});
