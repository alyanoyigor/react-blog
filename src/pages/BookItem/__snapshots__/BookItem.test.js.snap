// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem should match to snapshot with success response data 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1yd0tbh-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-97560v-MuiSvgIcon-root"
      data-testid="ArrowBackIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
      />
    </svg>
    <span>
      Back
    </span>
  </button>
  <div
    class="MuiBox-root css-gid8lu"
  >
    <h1
      class="MuiTypography-root MuiTypography-h4 css-ri2m7k-MuiTypography-root"
      data-testid="book-title"
    >
      1
    </h1>
    <p
      class="MuiTypography-root MuiTypography-subtitle1 css-118o8hd-MuiTypography-root"
    >
      Jul 25, 2022
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
    >
      1
       pages
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
    >
      1
    </p>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
    >
      1
    </p>
  </div>
</div>
`;
